ksql-plugins/ksql-rocksdb-config-setter/src/main/java/io/confluent/ksql/rocksdb/KsqlBoundedMemoryRocksDBConfig.java

```diff
@@ -0,0 +1,132 @@
+  private static final long BLOCK_SIZE = 4096L;
+
+  private static org.rocksdb.Cache cache;
+  private static org.rocksdb.WriteBufferManager writeBufferManager;
+
+  private static AtomicBoolean configured = new AtomicBoolean(false);
+  private static long totalOffHeapMemory;
+  private static long memtableSize;
+  private static long totalMemtableMemory;
+  private static int numBackgroundThreads;
+
+  public static void configure(final Map<String, Object> config) {
+    configure(config, new Options());
+  }
+
+  @VisibleForTesting
+  static void configure(final Map<String, Object> config, final Options options) {
+    if (configured.getAndSet(true)) {
+      throw new IllegalStateException(
+          "KsqlBoundedMemoryRocksDBConfig has already been configured. Cannot re-configure.");
+    }
+
+    try {
+      limitTotalMemory(config);
+      configureNumThreads(config, options);
+    } catch (IllegalArgumentException e) {
+      configured.set(false);
+      throw e;
+    }
+  }
+
+  @VisibleForTesting
+  static void reset() {
+    configured.set(false);
+  }
+
+  private static void limitTotalMemory(final Map<String, Object> config) {
+    if (!config.containsKey(TOTAL_OFF_HEAP_MEMORY_CONFIG)) {
+      throw new IllegalArgumentException(
+          "Missing configuration: " + TOTAL_OFF_HEAP_MEMORY_CONFIG);
+    }
+    totalOffHeapMemory = Long.parseLong((String)config.get(TOTAL_OFF_HEAP_MEMORY_CONFIG));
+
+    memtableSize = totalOffHeapMemory / 2 / N_MEMTABLES;
+    totalMemtableMemory = N_MEMTABLES * memtableSize;
+
+    cache = new org.rocksdb.LRUCache(totalOffHeapMemory, -1, false, INDEX_FILTER_BLOCK_RATIO);
+    writeBufferManager = new org.rocksdb.WriteBufferManager(totalMemtableMemory, cache);
+  }
+
+  private static void configureNumThreads(final Map<String, Object> config, final Options options) {
+    if (!config.containsKey(N_BACKGROUND_THREADS_CONFIG)) {
+      throw new IllegalArgumentException("Missing configuration: " + N_BACKGROUND_THREADS_CONFIG);
+    }
+    numBackgroundThreads = Integer.parseInt((String)config.get(N_BACKGROUND_THREADS_CONFIG));
+
+    // All Options share the same Env, and share a thread pool as a result
+    options.getEnv().setBackgroundThreads(numBackgroundThreads);
+  }
+
+  @Override
+  public void setConfig(
+      final String storeName,
+      final Options options,
+      final Map<String, Object> configs) {
+    if (!configured.get()) {
+      throw new IllegalStateException(
+          "Cannot use KsqlBoundedMemoryRocksDBConfig before it's been configured.");
+    }
+
+    final BlockBasedTableConfig tableConfig = (BlockBasedTableConfig)options.tableFormatConfig();
+
+    tableConfig.setBlockCache(cache);
+    tableConfig.setCacheIndexAndFilterBlocks(true);
+    options.setWriteBufferManager(writeBufferManager);
+
+    tableConfig.setCacheIndexAndFilterBlocksWithHighPriority(true);
+    tableConfig.setPinTopLevelIndexAndFilter(true);
+    tableConfig.setBlockSize(BLOCK_SIZE);
```

Comment:
You don't need to set this. The streams code should set this for you. This class should just be overrides.